core/domain/topic_domain.py

```diff
@@ -1192,6 +1193,28 @@ def validate(self, strict: bool = False) -> None:
                     raise utils.ValidationError(
                         'Subtopic with title %s does not have any skills '
                         'linked.' % subtopic.title)
+            subtopic_skill_ids.extend(subtopic.skill_ids)
+
+        skill_ids_that_are_not_in_subtopics = (
```

Comment:
Can you have this variable name contain "diagnostic test" somehow? Otherwise it feels like it may be referring to skills that are in the topic in general (since, when the topic is a work in progress, I think it's possible for skills to be uncategorized).

Speaking of which, you might need to check that uncategorized field as well.

Also, in future, if skills are added/deleted from the topic, make sure to update the diagnostic test field accordingly... (basically, think of the possible cases in the frontend and make sure to address them).

Reply:
> The skill_ids_for_diagnostic_test will only contain those skill_ids which are associated with a subtopic. Regarding the uncategorized skill, I am not able to think of any potential case when they relate, please help me find the connection between the two.

There's an uncategorized_skill_ids property in TopicModel. That relates to skills which are in the topic but which haven't been assigned to a subtopic yet. Such skills are valid diagnostic skill IDs so you'd need to account for them accordingly, e.g. in cases where the topic editor assigns the diagnostic test skills before assigning the skills to subtopics.

So I think the case to be careful about is when a skill is completely removed from a topic, not just from a subtopic (if it still remains in the topic after that). I agree that there is nothing you need to do when adding skills to the topic.

Reply:
Thank you @seanlip for the clarification, 
In the topic editor, I am thinking of providing options only for those skills, that are associated with a subtopic.
Eg: A skill X cannot be assigned to the diagnostic test if it is uncategorized in a topic or not assigned in the topic.

Also in the later moment of time when any skill gets assigned to a subtopic, then only the diagnostic test dropdown editor, will show it as an option for assigning that skill to the diagnostic test.

Looking forward to your point.
Thank you

Reply:
I think I prefer the case where any skill in the topic is eligible for the diagnostic test. That will prevent churn if the skill grouping within the topic is rearranged. We think of subtopics as a loose way to group the skills in the topic into more digestible chunks, rather than strong objects with permanent identity.

Reply:
P.S. Please don't forget to reassign me (or anyone) if you have comments that you would like them to answer (i.e. if the "next action" on the PR is theirs in any way).

Reply:
Ok, this means, that if an editor adds skill X into the diagnostic test from the uncategorized skill section they still are able to publish the topic, right?

Reply:
Yes, I think that's fine. If we wanted to be vigilant about this we should just ensure that published topics have no uncategorized skills.

Reply:
Updated.